alg/teca_temporal_reduction.py

```diff
@@ -0,0 +1,506 @@
+            return a pair of time steps bracketing the current month.
+            both returned time steps belong to the current month.
+            """
+            # check for more months to process
+            if (self.year > self.t1.year) or \
+                    (self.year == self.t1.year) and \
+                    (self.month > self.t1.month):
+                raise StopIteration
+
+            # find the time step of the first day
+            year = self.year
+            month = self.month
+
+            t0 = '%04d-%02d-01 00:00:00' % (self.year, self.month)
+            i0 = teca_py.coordinate_util.time_step_of(self.t, True, True,
+                                                      self.calendar,
+                                                      self.units, t0)
+
+            # find the time step of the last day
+            n_days = self.last_day_of_month()
+            t1 = '%04d-%02d-%02d 11:59:59' % (self.year, self.month, n_days)
+            i1 = teca_py.coordinate_util.time_step_of(self.t, True, True,
+                                                      self.calendar,
+                                                      self.units, t1)
+
+            # move to next month
+            self.month += 1
+
+            # move to next year
+            if self.month == 13:
+                self.month = 1
+                self.year += 1
+
+            return c_struct(time=self.t[i0], year=year, month=month,
+                            day=1, start_index=i0, end_index=i1)
+
+    class day_iterator:
+        """
+        An iterator over all days between 2 time_point's. A pair
+        of time steps bracketing the current day are returned at
+        each iteration.
+        """
+
+        def __init__(self, t, units, calendar):
+            """
+            t - an array of floating point time values
+            units - string units of the time values
+            calendar - string name of the calendar system
+            """
+            # time values
+            self.t = t
+            self.units = units
+
+            calendar = calendar.lower()
+            self.calendar = calendar
+
+            self.noleap = \
+                True if (calendar == 'noleap') or (calendar == 'no_leap') \
+                or (calendar == '365_day') or (calendar == '360_day') else \
+                False
+
+            # time point's to iterate between
+            self.t0 = time_point(t[0], units, calendar)
+            self.t1 = time_point(t[-1], units, calendar)
+
+            # current time state
+            self.year = self.t0.year
+            self.month = self.t0.month
+            self.day = self.t0.day
+
+        def last_day_of_month(self):
+            """
+            get the number of days in the month, with logic for
+            leap years
+            """
+
+            # first entry is -1 because month numbers start at 1 not zero
+            days = [-1, 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
```

Comment:
same comment here as above: can we use `ccs_dpm()` inside this function instead?